include/pagmo/algorithms/cstrs_self_adaptive.hpp

```diff
@@ -0,0 +1,783 @@
+                    // 1 - Every 50 lines print the column names
+                    if (count % 50u == 1u) {
+                        print("\n", std::setw(7), "Iter:", std::setw(15), "Fevals:", std::setw(15), "Best:",
+                              std::setw(15), "Infeasibility:", std::setw(15), "Violated:", std::setw(15), "Viol. Norm:",
+                              std::setw(15), "N. Feasible:", '\n');
+                    }
+                    // 2 - Print
+                    auto cur_best_f = pop.get_f()[pop.best_idx()];
+                    auto c1eq = detail::test_eq_constraints(cur_best_f.data() + 1, cur_best_f.data() + 1 + nec,
+                                                            prob.get_c_tol().data());
+                    auto c1ineq = detail::test_ineq_constraints(cur_best_f.data() + 1 + nec,
+                                                                cur_best_f.data() + cur_best_f.size(),
+                                                                prob.get_c_tol().data() + nec);
+                    auto n = prob.get_nc() - c1eq.first - c1ineq.first;
+                    auto l = c1eq.second + c1ineq.second;
+                    auto infeas = penalized_udp_ptr->compute_infeasibility(penalized_udp_ptr->m_f_hat_down);
+                    auto n_feasible = penalized_udp_ptr->m_n_feasible;
+                    print(std::setw(7), iter, std::setw(15), prob.get_fevals() - fevals0, std::setw(15), cur_best_f[0],
+                          std::setw(15), infeas, std::setw(15), n, std::setw(15), l, std::setw(15), n_feasible);
+                    if (!prob.feasibility_f(pop.get_f()[pop.best_idx()])) {
+                        std::cout << " i";
+                    }
+                    ++count;
+                    std::cout << std::endl; // we flush here as we want the user to read in real time ...
+                    // Logs
+                    m_log.emplace_back(iter, prob.get_fevals() - fevals0, cur_best_f[0], infeas, n, l, n_feasible);
+                }
+            }
+            // We call the evolution on the unconstrained population (here is where fevals will increase)
+            new_pop = static_cast<const algorithm *>(this)->evolve(new_pop);
+            penalized_udp_ptr = new_pop.get_problem().extract<detail::penalized_udp>();
+            // We update the original pop avoiding fevals thanks to the cache
+            for (decltype(pop.size()) i = 0u; i < pop.size(); ++i) {
+                auto x = new_pop.get_x()[i];
+                auto it_f = penalized_udp_ptr->m_fitness_map.find(x);
+                if (it_f != penalized_udp_ptr->m_fitness_map.end()) { // cash hit
+                    pop.set_xf(i, x, it_f->second);
+                } else { // we have to compute the fitness (this will increase the feval counter in the ref pop problem,
+                         // but should never happen hence we do not test it)
+                    pop.set_x(i, x); // LCOV_EXCL_LINE
```

Comment:
1) Are we 100% sure the cahce as its implemented now will never fail? (before we had some cases where it could have returned unccorrectly the absence of a dv)
2) Are we 100% sure that we are never calling this on chromosomes previously unseen? (I think so, but did not want to bother to form in my mind a formal guarantee)

Anyway, will remove.